index: Commit MuHash and best block together for coinstatsindex

mzumsande opened the pull request:
Fixes #24076

Coinstatsindex currently writes the MuHash (`DB_MUHASH`) to disk in `CoinStatsIndex::WriteBlock()` and `CoinStatsIndex::ReverseBlock()`, but the best synced block is written in `BaseIndex::Commit()`. These are called at different points in time, both during the ThreadSync phase, and also after the initial sync is finished and validation callbacks (`BlockConnected()` vs `ChainStateFlushed()`) perform the syncing.

As a result, the index DB is temporarily in an inconsistent state, and if bitcoind is terminated uncleanly (so that there is no time to call `Commit()` by receiving an interrupt or by flushing the chainstate) this leads to problems:
On the next startup, `Init()` will read the best block and a MuHash that corresponds to a different (higher) block. Indexing will  be picked up at the the best block processing some blocks again, but since MuHash is a rolling hash, it will process some utxos twice and the muhashes for all future blocks will be wrong, as was observed in #24076.

Fix this by always committing `DB_MUHASH` together with `DB_BEST_BLOCK`.

Note that the block data for the index is still written at different times, but this does not corrupt the index - at worst, these entries will be processed another time and overwritten after an unclean shutdown and restart.

mzumsande committed:
index: Commit DB_MUHASH and DB_BEST_BLOCK to disk together

If these are written to disk at different times,
unclean shutdowns can lead to index corruption.

mzumsande commented:
FYI @fjahr

luke-jr commented on src/index/coinstatsindex.cpp:
```diff
@@ -481,5 +486,5 @@ bool CoinStatsIndex::ReverseBlock(const CBlock& block, const CBlockIndex* pindex
     Assert(m_total_unspendables_scripts == read_out.second.total_unspendables_scripts);
     Assert(m_total_unspendables_unclaimed_rewards == read_out.second.total_unspendables_unclaimed_rewards);
 
-    return m_db->Write(DB_MUHASH, m_muhash);
```
Seems like the only problem is this lone line?

Sjors replied:
If I replace only that line with `true` the reorg test in `feature_coinstatsindex.py` still passes. But not sure if that helps with the original issue.

Sjors replied:
>  if the node gets shut down during the reorg,

I guess this is very non trivial to test, but maybe add a comment?

mzumsande replied:
I added a comment at the end of `CoinStatsIndex::WriteBlock` (as suggested by ryanofsky below). I think that duplicating this comment in `CoinStatsIndex::ReverseBlock` is not required, since this function doesn't write anything else to disk there wouldn't really be the expectation that for it to write `DB_MUHASH` here.

mzumsande commented on src/index/coinstatsindex.cpp:
```diff
@@ -228,10 +228,7 @@ bool CoinStatsIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
     m_muhash.Finalize(out);
     value.second.muhash = out;
 
-    CDBBatch batch(*m_db);
-    batch.Write(DBHeightKey(pindex->nHeight), value);
-    batch.Write(DB_MUHASH, m_muhash);
-    return m_db->WriteBatch(batch);
```
I think this is spot is problematic, and the reason for the behavior reported in #24076: 

If we persist `DB_MUHASH` here in each `BlockConnected()` callback, but not `DB_BEST_BLOCK` (which is persisted only in `ChainStateFlushed()` callbacks and will therefore at this point in time have a lower height), the db is temporarily in an inconsistent state.

So in case of an unclean shutdown (with no time to flush the chainstate to disk to get in sync again), on next startup the index will initialize with the older `DB_BEST_BLOCK` but a `DB_MUHASH` which contains data from newer blocks. The index then reprocesses the blocks after `DB_BEST_BLOCK`'s height a second time. 
But MuHash is a rolling hash, so it will incorporate data from these blocks twice, and the hash will be incorrect.

ryanofsky replied:
In commit "index: Commit DB_MUHASH and DB_BEST_BLOCK to disk together" (010156c07f28e2c8dc30d4408942d1e1d88ab54d)

> The `CommitInternal` override feels like a hacky refactor.

I think I agree with luke if the point is that the override-and-call-base class code in `CommitInternal` is a little hacky, but fjahr's right this just the way the base index class was designed, so the override seems like appropriate way to handle this now. #24230 does clean up all the overrides and will simplify this a bit.

mzumsande replied:
Just to make this clear: This is not a refactor, it changes behavior (the points in time when `DB_MUHASH` is written to disk).

My understanding is that the indexes work in general such that that all persisted data describing the current _progress_ of indexing (such as the best block) must be in sync with the persisted chainstate information (which is sometimes not flushed for several blocks received) - that is why `ChainStateFlushed()` callbacks are used. In case of an unclean shutdown, the chainstate will be constructed from this point on, so the indexes must as well.

In contrast to this is the actual indexed data for each block - I think we could in principle cache it and also write it to disk with each `ChainStateFlushed()` callback and not use `BlockConnected()` at all. 

But it is also possible to write it to disk in advance with each `BlockConnected()` callback, as it is done:
This is dependent on the assumption that we will not read (and possibly overwrite) the indexed data in case it is ahead of the flushed chainstate data (and the best block), and also that _before_ flushing the chainstate, we have called `BlockConnected()` for each block.

With this in mind, I don't see a better way of doing this other than overriding `CommitInternal`.

Sjors reviewed (commented):
Concept ACK

It might be worth rebasing this on #24133 and if at all possible, add a test that triggers the error added there.

I think overriding `CommitInternal` makes sense given that we require `DB_MUHASH` and `DB_BEST_BLOCK` to be in sync (which we explicitly check as of #24133).

The alternative would be to drop `DB_MUHASH` altogether, but since MuHash can't (?) be initialized from a finalized hash, we would have to index both the denominator and numerator.

fjahr reviewed (commented):
Code review ACK 010156c07f28e2c8dc30d4408942d1e1d88ab54d

While I can not reproduce the issue I agree that this is likely to fix #24076.

I think there could still be inconsistencies between the running muhash and the coinstats data since those are not batched anymore now but at least that would then lead to a failure in `Init()` at restart. This could be avoided by squashing all of that into `CommitInternal()` as well ([drafted here](https://github.com/fjahr/bitcoin/commit/a90394efb9b348b7d3615d61fbea612470b1c75c)) but I don't think this is necessary right now. As I have reviewed this I am now really not sure about the boundaries between `Commit()` and `WriteBlock()` anymore and will think about refactoring that while reviewing #24230 (maybe that's already solved there).

ryanofsky commented on src/index/coinstatsindex.cpp:
```diff
@@ -228,10 +228,7 @@ bool CoinStatsIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
     m_muhash.Finalize(out);
     value.second.muhash = out;
 
-    CDBBatch batch(*m_db);
```
In commit "index: Commit DB_MUHASH and DB_BEST_BLOCK to disk together" (010156c07f28e2c8dc30d4408942d1e1d88ab54d)

I might add a code comment here like "Intentionally do not update DB_MUHASH here so it stays in sync with DB_BEST_BLOCK, and index is not corrupted if there is an unclean shutdown"

mzumsande replied:
Thanks, I added a comment here.

ryanofsky reviewed (approved):
Code review ACK 010156c07f28e2c8dc30d4408942d1e1d88ab54d

It would be good to add a unit test to coinstatsindex_tests that triggers the issue this is supposed to fix. I think the test would just need to make the index object, send a BlockConnected notification, and then destroy the index object without sending ChainStateFlushed notification. This would save a corrupted index that should result in errors next time it is loaded.


re: https://github.com/bitcoin/bitcoin/pull/24138#pullrequestreview-880885450

> I think there could still be inconsistencies between the running muhash and the coinstats data since those are not batched anymore now but at least that would then lead to a failure in `Init()` at restart.

I could be wrong but this seems fine to me since coinstats data is accessed by height and written before the muhash & best block. So if the coinstats data is ahead of the muhash. it is just ignored. In your suggested followup a90394efb9b348b7d3615d61fbea612470b1c75c, I think the `if (!m_db->Read(DBHeightKey(pindex->nHeight))` check would be good to add, but I think it should it should just trigger a fatal error if the key isn't found. I don't see how the recovery code added in that commit would work.

> #24230 (maybe that's already solved there).

My guess is #24230 doesn't affect this, since it's trying to leave index's internal behavior the same, and just change the way it receives block connected and disconnected notifications, without changing the way it handles them.

mzumsande commented:
I pushed an update, rebasing (with #24133 merged) and addressing feedback (adding a comment).

> It would be good to add a unit test to coinstatsindex_tests that triggers the issue this is supposed to fix. I think the test would just need to make the index object, send a BlockConnected notification, and then destroy the index object without sending ChainStateFlushed notification. This would save a corrupted index that should result in errors next time it is loaded.

This would be great. I haven't done this with this push, because 
a) I couldn't make it work yet, I had some technical difficulties with the test framework when restarting the index a second time. I will look into it more unless
b) @fjahr who said [here](https://github.com/bitcoin/bitcoin/pull/24133#discussion_r810605240) that he is working on this has more success.

ryanofsky committed:
Add coinstatsindex_unclean_shutdown test

ryanofsky reviewed (approved):
Code review ACK eb6cc05da32c5bde122725a0bc907d3767a791cd.  Only changes since previous review are rebasing and adding comments.

@mzumsande I wrote a little test which passes with this bugfix, and fails with the fix reverted. Feel free to add it to this PR or use the code in any way: ~ded9fda2b6e154ac64b2fc755ddaa738b309b7e9~ d3c2ebbfdc3c61a04b645c6bb435424fa60f4a0b

mzumsande commented:
> @mzumsande I wrote a little test which passes with this bugfix, and fails with the fix reverted. Feel free to add it to this PR or use the code in any way: ~[ded9fda](https://github.com/bitcoin/bitcoin/commit/ded9fda2b6e154ac64b2fc755ddaa738b309b7e9)~ [d3c2ebb](https://github.com/bitcoin/bitcoin/commit/d3c2ebbfdc3c61a04b645c6bb435424fa60f4a0b)

Thank you! The test looks great, I added your commit to this PR (and confirmed that it fails on master).

ryanofsky reviewed (approved):
Code review ACK 691d45fdc83ec14f87a400f548553168ac70263f. Only change since last review is adding test

fjahr reviewed (commented):
ACK 691d45fdc83ec14f87a400f548553168ac70263f

Reviewed code and ensured that the newly added test fails when eb6cc05da32c5bde122725a0bc907d3767a791cd is reverted.

MarcoFalke commented on src/test/coinstatsindex_tests.cpp:
```diff
@@ -78,4 +85,44 @@ BOOST_FIXTURE_TEST_CASE(coinstatsindex_initial_sync, TestChain100Setup)
+    {
+        CoinStatsIndex index{1 << 20};
+        // Make sure the index can be loaded.
+        BOOST_REQUIRE(index.Start(chainstate));
```
Is there a huge cost to spinning up the second index? If not, I think both versions are fine.
